sw/device/lib/base/csr.h

```diff
@@ -0,0 +1,2786 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/stdasm.h"
+
+// Define function implementations (mocked or real).
+#ifdef MOCK_CSR
+
+uint32_t mock_csr_read(const char *name);
+#define CSR_READ_IMPL(name, addr) return mock_csr_read(#name);
+
+uint32_t mock_csr_write(const char *name, uint32_t value);
+#define CSR_WRITE_IMPL(name, addr) return mock_csr_write(#name, value);
+
+uint32_t mock_csr_set_bits(const char *name, uint32_t mask);
+#define CSR_SET_BITS_IMPL(name, addr) return mock_csr_set_bits(#name, mask);
+
+uint32_t mock_csr_clear_bits(const char *name, uint32_t mask);
+#define CSR_CLEAR_BITS_IMPL(name, addr) return mock_csr_clear_bits(#name, mask);
+
+#else  // MOCK_CSR
+
+#define CSR_READ_IMPL(name, addr)                  \
+  uint32_t old;                                    \
+  asm volatile("csrr %0, " #addr ";" : "=r"(old)); \
+  return old;
+
+#define CSR_WRITE_IMPL(name, addr)                                   \
+  uint32_t old;                                                      \
+  asm volatile("csrrw %0, " #addr ", %1;" : "=r"(old) : "r"(value)); \
+  return old;
+
+#define CSR_SET_BITS_IMPL(name, addr)                               \
+  uint32_t old;                                                     \
+  asm volatile("csrrs %0, " #addr ", %1;" : "=r"(old) : "r"(mask)); \
+  return old;
+
+#define CSR_CLEAR_BITS_IMPL(name, addr)                             \
+  uint32_t old;                                                     \
+  asm volatile("csrrc %0, " #addr ", %1;" : "=r"(old) : "r"(mask)); \
+  return old;
+
+#endif  // MOCK_CSR
+
```

Comment:
I think it would be nice to split out the functionality that is above this line into its own file. Anything that is below would then use the above file.

Reply:
Maybe even into multiple with routines logically grouped within them.